### ISSUE TITLE
Random SIGPIPE on ipc handle

### ISSUE THREAD

**Issue by boryspoplawski** [oscarlab/graphene#1534]
## Description of the problem
Sometimes a process will receive a SIGPIPE on ipc handle which will cause it to shutdown.

## Steps to reproduce
In ltp tests:
`i=0; while [[ $i == 0 ]]; do ./pal_loader ./fork11 ; i=$?; done`

## Expected results
All runs succeed.

## Actual results
After some time it will stop with:
```fork11      2  TBROK  :  tst_sig.c:234: unexpected signal SIGPIPE(13) received (pid = 1).```

**Comment by dimakuv** [comment]
I'm pretty sure it's because of this PR: https://github.com/oscarlab/graphene/pull/1495.

Basically, we merged correct emulation of SIGPIPE now, which broke some other part of Graphene (most probably the infamous race on "Linux kills the process" vs "Graphene sends/receives the SIGCHLD signal").

**Comment by boryspoplawski** [comment]
Seems like that's it. We need to distinguish SIGPIPE on graphene internal file descriptors vs app fds, handle the former and pass the latter to the app.

**Comment by dimakuv** [comment]
Ok, so the bug happens because Graphene has internal pipes for IPC communication. In particular, `send_ipc_message()` invokes `DkStreamWrite()` on an IPC pipe between the parent and the child, and the read end of the pipe may be already closed because the child exited. So since Graphene supports SIGPIPE now and propagates it from host to LibOS to app (https://github.com/oscarlab/graphene/pull/1495), it ends up triggering a SIGPIPE handler, which is unexpected for the application. For example, LTP's `fork11` registers a SIGPIPE handler just to test that SIGPIPE is actually never raised (and thus fails).

**Proposed solution**. The simplest way is to emulate SIGPIPE completely inside LibOS. For this:
1. PALs set `sigaction(SIGPIPE, SIG_IGN)` to ignore SIGPIPEs from the host. This forces Linux to return from writes on pipes/UNIX sockets with `-EPIPE`.
2. These `-EPIPE`s will propagate to the LibOS at the places of `DkStreamWrite()`. In particular, emulations of pipes and UNIX sockets must be augmented with a special case of `-EPIPE`: if the app registered its own signal handler, this `-EPIPE` must trigger a LibOS-internal signal delivery to the whole process (SIGPIPE is process-wide). It's better to merge https://github.com/oscarlab/graphene/pull/1510 before this change.
3. At the internal-graphene-pipe snippets in IPC code, `-EPIPE` will *not* trigger a signal delivery but instead will keep the old logic.

This solution effectively undoes https://github.com/oscarlab/graphene/pull/1495. The positive is that Graphene gets rid of the complicated semantics of SIGPIPE and becomes generally more portable.

**Comment by stefanberger** [comment]
Let me give this a shot via the test application I added to LibOS. The above test case works fine on my 4 vCPU system literally forever - fork11 always passes.

**Comment by stefanberger** [comment]
Before #1495 Pal had a signal handler for SIGPIPE but was more or less ignoring the signal, with #1495 we are not forwarding all signals to LibOS. If there was a way to distinguish between internal file descriptors and LibOS level ones, we could keep the old behavior for internal ones and invoke LibOS for those on the LibOS level. I suppose there's not way for use to look up a file descriptor and determine it's for LibOS.

Looking at it at the LibOS level, the call to internal_fault() could be the reason for the fault. Assuming the check for is_internal_tid() works correctly, we should probably comment out the else branch:

```
static void pipe_upcall(PAL_PTR event, PAL_NUM arg, PAL_CONTEXT* context) {
    if (!is_internal_tid(get_cur_tid()))
        deliver_signal(ALLOC_SIGINFO(SIGPIPE, 0, si_pid, 0), /*context=*/NULL);
#if 0
    else
        internal_fault("Internal SIGPIPE fault", arg, context);
#endif
    DkExceptionReturn(event);
}
```

Unfortunately I cannot test this with the endless loop on `fork11`.

**Comment by stefanberger** [comment]
Scratch this. I can test it and it's not the solution.

**Comment by stefanberger** [comment]
PR #1543 fixes the `fork11` issue and the regression test case also works.

**Comment by boryspoplawski** [comment]
@stefanberger we already had a PR (#1542, it's marked as closing this issue above) that deals with this issue ...

**Comment by stefanberger** [comment]
@boryspoplawski oh dear... hadn't looked.

**Comment by stefanberger** [comment]
... looks a lot more complicated.

**Comment by dimakuv** [comment]
@stefanberger Yep, we raced on fixing this issue :)

My PR looks more complicated for two reasons: (1) it is built on top of Borys's PR which constitute the bulk of changed lines -- it will be rebased once Borys's PR is merged, and (2) I refactored `db_exception.c` logic in both Linux and Linux-SGX PALs as part of this work.

If you look at only my 3-4 last commits, you'll see that our solutions are very much similar. I'd prefer to go with my PR since it also refactors stuff and is more explicit about the shim handles which receive special SIGPIPE treatment. But I'll be happy with your comments!

Also, sorry for this duplicated effort. I assigned this issue to myself on GitHub (see the right panel) a couple days ago; we typically do not take on issues if they were assigned to someone else. (Not that it's a problem, just unnecessarily duplicates efforts.)
